### PR TITLE
port ZBar recipe to conan2

### PR DESCRIPTION
**Pull request by jdoubleu**
Specify library name and version:  **zbar/0.23.92**

---

- [x] I've read the [contributing guidelines](https://github.com/conan-io/conan-center-index/blob/master/CONTRIBUTING.md).
- [x] I've used a [recent](https://github.com/conan-io/conan/releases/latest) Conan client version close to the [currently deployed](https://github.com/conan-io/conan-center-index/blob/master/.c3i/config_v1.yml#L6).
- [x] I've tried at least one configuration locally with the [conan-center hook](https://github.com/conan-io/hooks.git) activated.

### CODE REVIEW

**Comment by prince-chrismc** [comment]
Make sure to comment in #4  for access

Also this is missing the test_v1_package which you can find in the docs :)

File: recipes/zbar/all/conanfile.py [hunk 1]
```diff
@@ -1,7 +1,11 @@
-from conans import tools, AutoToolsBuildEnvironment
-from conans.errors import ConanInvalidConfiguration
 from conan import ConanFile
+from conan.errors import ConanInvalidConfiguration
+from conan.tools.apple import is_apple_os
 from conan.tools.build import cross_building
+from conan.tools.files import get, copy, chdir, rmdir, rm, collect_libs
```
Comment (prince-chrismc) [inline recipes/zbar/all/conanfile.py:5]:
```suggestion
from conan.tools.files import get, copy, rmdir, rm, collect_libs
```

File: recipes/zbar/all/conanfile.py [hunk 2]
```diff
@@ -1,7 +1,11 @@
...
+from conan.tools.gnu import Autotools, AutotoolsToolchain
+from conan.tools.layout import basic_layout
```
Comment (prince-chrismc) [inline recipes/zbar/all/conanfile.py:7]:
```suggestion
```

File: recipes/zbar/all/conanfile.py [hunk 3]
```diff
@@ -1,7 +1,11 @@
...
+from conan.tools.scm import Version
 import os
 import shutil
```
Comment (prince-chrismc) [inline recipes/zbar/all/conanfile.py:10]:
```suggestion
```

**Review by prince-chrismc — COMMENTED** [review]
Clean imports (linter comments)

File: recipes/zbar/all/conanfile.py [hunk 4]
```diff
@@ -87,23 +85,25 @@ def build_requirements(self):
...
+        tc = AutotoolsToolchain(self)
+        tc.generate()
 
     def _configure_autotools(self):
         if not self._autotools:
-            if tools.Version(self.version) >= "0.22":
-                with tools.chdir(self._source_subfolder):
-                    self.run("autoreconf -fiv")
-            self._autotools = AutoToolsBuildEnvironment(self)
+            self._autotools = Autotools(self)
```
Comment (prince-chrismc) [inline recipes/zbar/all/conanfile.py:104]: You should no longer cache the the build helpers, you can check the package template for more details

**Review by prince-chrismc — COMMENTED** [review]
Overall this looks really good,

Two small things 

- the builder dont need to be cached anymore 
- there should be a test_v1_package

**Comment by jdoubleu** [comment]
Thanks for the help!

I'm not that familiar with Conan. I tried to fix it so that it just works™ with v2, which it did. I'm sure there's room for improvements.

> there should be a test_v1_package

Is the new file backwards compatible? Can it run with both Conan v1 and v2 simultaneously?

**Comment by prince-chrismc** [comment]
It's specifically talored to ConanCenter's build infrastructure and wont be run by default with `conan create` :)

It only works with Conan 1.x but needs to be manully run (see the docs "developing locally" if you want more details)

**Comment by prince-chrismc** [comment]
Also you need to comment in #4  for access

File: recipes/zbar/all/conanfile.py (hunk 4 above)
Comment (prince-chrismc) [reply recipes/zbar/all/conanfile.py:104]: You can check the template for the updates https://github.com/conan-io/conan-center-index/blob/7ce55ac4cb6d1f53ffaca6bb767e6f25655c4b67/docs/package_templates/autotools_package/all/conanfile.py#L150

:) Thanks for updating this recipe

**Comment by prince-chrismc** [comment]
Make sure to comment in #4

**Comment by jdoubleu** [comment]
Thanks! Does it address all the previous comments in this PR?
Sorry, I haven't had time to fix it myself.

**Comment by uilianries** [comment]
> Thanks! Does it address all the previous comments in this PR?

Yes, all them 😄 

Autotools is not happy with Clang 😤

I'll take a look this afternoon.

File: recipes/zbar/all/conanfile.py [hunk 5]
```diff
@@ -87,69 +85,61 @@ def build_requirements(self):
...
-        shutil.copy(self._user_info_build["gnu-config"].CONFIG_SUB,
-                    os.path.join(self._source_subfolder, "config", "config.sub"))
-        shutil.copy(self._user_info_build["gnu-config"].CONFIG_GUESS,
-                    os.path.join(self._source_subfolder, "config", "config.guess"))
+        copy(self, "config.sub", src=self.source_folder, dst=os.path.join(self.source_folder, "config"))
+        copy(self, "config.guess", src=self.source_folder, dst=os.path.join(self.source_folder, "config"))
 
-        autotools = self._configure_autotools()
+        autotools = Autotools(self)
+        autotools.autoreconf()
```
Comment (uilianries) [inline recipes/zbar/all/conanfile.py:126]:
```suggestion
```
The `configure` file is already generated and re-running autoconf generates more warnings.

File: recipes/zbar/all/conanfile.py [hunk 6]
```diff
@@ -87,69 +85,60 @@ def build_requirements(self):
...
     def build(self):
-        shutil.copy(self._user_info_build["gnu-config"].CONFIG_SUB,
-                    os.path.join(self._source_subfolder, "config", "config.sub"))
-        shutil.copy(self._user_info_build["gnu-config"].CONFIG_GUESS,
-                    os.path.join(self._source_subfolder, "config", "config.guess"))
+        copy(self, "config.sub", src=self.source_folder, dst=os.path.join(self.source_folder, "config"))
+        copy(self, "config.guess", src=self.source_folder, dst=os.path.join(self.source_folder, "config"))
 
-        autotools = self._configure_autotools()
+        autotools = Autotools(self)
```
Comment (uilianries) [inline recipes/zbar/all/conanfile.py:125]:
```suggestion
        autotools = Autotools(self)
        if Version(self.version) > "0.10":
            autotools.autoreconf()
```
The version 0.23 is available on Github, but there is no configure file available.

**Comment by uilianries** [comment]
@jdoubleu Hello, please, consider the new PR https://github.com/jdoubleu/conan-center-index/pull/2. The versions 0.10 and 0.23 are basically different in terms of building script, so need to patch the 0.10 to be compatible with Mac M1.

**Comment by uilianries** [comment]
@jdoubleu Finally is working! 🔥

**Review by uilianries — APPROVED** [review]

**Review by SSE4 — APPROVED** [review]

**Pull request merged**